### ISSUE TITLE
Python 3.6 and 3.7 support [Feature Request]

### ISSUE THREAD

**Issue by leycec** [The-Compiler/pytest-xvfb#20]
It'd be juicy if `pytest-xvfb` officially supported modern Python releases.

Python 3.6 is [nearly three years old](https://www.python.org/downloads/release/python-360) as of this writing, Python 3.7 is [no spring chicken either](https://www.python.org/downloads/release/python-370), and Python 3.8 is [waiting in the wings](https://www.python.org/dev/peps/pep-0569). The river of time passes swiftly indeed for all good Python projects.

## But What Can Possibly Be Done?

To rectify this mounting tragedy, a modest checklist of mandatory work might resemble:

- [ ] Locally validate `pytest-xvfb` to operate as expected under Python 3.6.
  - [ ] If not, submit pull request(s) rectifying this.
- [ ] Locally validate `pytest-xvfb` to operate as expected under Python 3.7.
  - [ ] If not, submit pull request(s) rectifying this.
- [ ] Add Python 3.6 and 3.7 to the:
  - [ ] `tox.ini` build matrix.
  - [ ] `.travis.yml` build matrix.
  - [ ] `appveyor.yml` build matrix.
  - [ ] Classifiers listed by `setup.py`.
- [ ] Release a new stable version – say, `pytest-xvfb` 1.3.0?

## But What About Downstream Dependencies?

**Right.** This is a known conundrum.

The most recent stable release of PyVirtualDisplay also fails to support Python 3.6 and 3.7, which admittedly complicates matters. Fortunately, the live version of PyVirtualDisplay added explicit support for Python 3.6 [several years ago](https://github.com/ponty/PyVirtualDisplay/commit/dff144f26670dc193fbb95a26961372d5b6c765b). Unfortunately, these changes have yet to be formally released into a new stable version of PyVirtualDisplay. Fortunately, I [recently submitted an issue](https://github.com/ponty/PyVirtualDisplay/issues/37) requesting that @ponty kindly resolve these complaints by adding explicit support for Python 3.7 and then releasing a new stable version. Unfortunately, @ponty no longer appears to be maintaining PyVirtualDisplay. A community fork may be desirable, at some point.

Until then, locally testing `pytest-xvfb` against the live version of PyVirtualDisplay under Python 3.6 would at least enable `pytest-xvfb` developers to add support for Python 3.6. `</shrug>`

## But Are You (@leycec) Actually Gonna Do Anything?

**Maybe.** Although available free time is limited, I *may* be able to lend a volunteer hand here.

Our [PySide2-based multiphysics biology simulator](https://gitlab.com/betse/betsee) requires both `pytest-xvfb` *and* [`pytest-qt`](https://github.com/pytest-dev/pytest-qt) to sanely automate continuous integration testing. Since we support Python 3.5 through 3.7, generalizing PyVirtualDisplay and `pytest-xvfb` to support modern Python versions is a hard prerequisite for that.

I'd like to first see whether @ponty replies to ponty/PyVirtualDisplay#37 or not. If PyVirtualDisplay genuinely *is* dead, that would chuck a monkey wrench into the well-laid plans of mice and men.

In conclusion, [motivational lyrics](https://www.youtube.com/watch?v=ZD_H1ePLylA) for 80's-era posterity:

> Time – what is time?
> I wish I knew how to tell you why.
> It hurts to know.
> Aren't we machines?
> Time – what is time?
> Unlock the door
> And see the truth;
> Then time is time again.

**Comment by The-Compiler** [comment]
You sure have a way with words! :+1: If only everyone put as much wit in their issues, maintaining open source would be much more fun!

I'm using pytest-xvfb in [qutebrowser](https://www.qutebrowser.org/) with Python 3.6 and 3.7 without any issues - though you're right that it should be updated to officially support those as well.

I'll get back to it in July, unless someone beats me to it. Currently focusing on my bachelor thesis! If you don't hear back by then, it might've gotten lost in my todo list (read: hundreds of mails marked as "important" in my mailbox), feel free to ping me again if that's the case!

**Comment by leycec** [comment]
> Currently focusing on my bachelor thesis!

This probably goes without saying, but... **please do that first.**

Of course, you'll have no problems whatsoever. Based on your [eclectic menagerie of popular projects](https://github.com/The-Compiler) (especially [qutebrowser](https://github.com/qutebrowser/qutebrowser) – a doctoral thesis on correct UX browser design if ever I've seen one), you're a veritable shoo-in. :mans_shoe: 

> If only everyone put as much wit in their issues...

*Oh, you.* :blush: 

> ...maintaining open source would be much more fun!

This saddens the pitchfork-wielding zealot in me. Having devoted my life to unpaid volunteerism for the past decade or so, I can only offer this tangential advice: "Life really *is* too short."

The lidded iris of time only dilates faster and faster as you age. One day, you're putting the finishing perfectionist touches on your incendiary bachelor's thesis; the next, you're graduating Summa Cum Laude with a dual degree in Molecular Gastronomy and Computational Ergometry. Suddenly, before you know it, you're a grizzled 39-year-old post-doctoral fellow discretely guzzling imported Polish vodka from a crinkled paper bag in [Underpass Park](http://www.explorewaterfrontoronto.ca/project/underpass-park). *This can happen to you.*

I suppose what I'm trying to say is this: when you no longer enjoy something (even if that something is the Eighth Wonder of open-source software development), it may be time to consider. Is that something genuinely worth your vanishing lifeblood?

In my experience as a grizzled 39-year-old hipster neckbeard, the answer is usually "Nope."

> I'm using pytest-xvfb in qutebrowser with Python 3.6 and 3.7 without any issues...

**Excellent.** This was my working assumption, but it's gratifying to hear that this has been stress-tested in a real-world production environment. As I'm currently adding a `pytest-xvfb` ebuild to my [third-party Gentoo Linux overlay](https://github.com/leycec/raiagent), I'll note that `pytest-xvfb` supports effectively everything. *Huzzah.*

Relatedly, @ponty of [PyVirtualDisplay fame](https://github.com/ponty/PyVirtualDisplay) recently added [official support for Python 3.6 and 3.7](https://github.com/ponty/PyVirtualDisplay/commit/254fac856ced019238cb0c1a32a3918eb8f64b31) as well. So, pytest-xvfb is now good to go as far as upstream dependencies are concerned. (*Thanks again, @ponty!*)

> I'll get back to it in July...

**Excellent ✕ 2.** As an inherently ~~lazy~~ patient man, I choose to wait this one out.

Thanks for the prompt and perspicacious reply, @The-Compiler. May the Summer swelter bring good things to all of us. Best of luck with your thesis! <sup>*...not that you need it, of course.*</sup>

**Comment by leycec** [comment]
**Good news!** You'll be getting a bachelor's. Also, Gentoo users may now easily install the full `pytest-xvfb` stack (i.e., `pytest-xvfb` **+** `pytest` **+** `Xvfb` **+** `PyVirtualDisplay` **+** `EasyProcess`) as follows:

```
eselect repository enable raiagent
emerge --sync raiagent
emerge --ask --autounmask pytest-xvfb
```

*Only the best for mah Gentoo peeps.*

**Comment by The-Compiler** [comment]
Incredible to see it's been a year (almost exactly!) since your awesome musings - but they haven't been forgotten!

I pushed a commit which *should* hopefully address everything - tests are still failing because of #22, but I'll take care of that next, and then release v2.0.0.

Incidentally, my SO just woke up from her post-term-project-deadline slumber, so I'll shift my priorities and save the remaining work for another day! :wink:
